component-runtime-impl/src/main/java/org/talend/sdk/component/runtime/input/InputImpl.java

```diff
@@ -56,21 +72,78 @@ protected InputImpl() {
         // no-op
     }
 
+    @Override
+    public void start() {
+        super.start();
+        initCheckpointFunctions();
+    }
+
+    @Override
+    public void start(final Consumer checkpointConsumer) {
+        start();
+        if (checkpointEnabled) {
+            checkpointCallback = checkpointConsumer;
+        }
+    }
+
+    protected void initCheckpointFunctions() {
+        if (checkpointEnabled) {
+            checkpoint = findMethods(CheckpointData.class).findFirst().orElse(null);
+            shouldCheckpoint = findMethods(CheckpointAvailable.class).findFirst().orElse(null);
+        }
+    }
+
     @Override
     public Object next() {
         if (next == null) {
             init();
         }
-        final Object record = readNext();
-        if (record == null) {
+        final Object data = readNext();
+        if (data == null) {
             return null;
         }
-        final Class<?> recordClass = record.getClass();
+        // do we need to checkpoint here?
+        if (isCheckpointReady() && checkpointCallback != null) {
+            checkpointCallback.accept(getCheckpoint());
+        }
+        final Class<?> recordClass = data.getClass();
         if (recordClass.isPrimitive() || String.class == recordClass) {
             // mainly for tests, can be dropped while build is green
-            return record;
+            return data;
+        }
+        return converters.toRecord(registry, data, this::jsonb, this::recordBuilderFactory);
+    }
+
+    @Override
+    public CheckpointState getCheckpoint() {
+        if (checkpoint != null) {
+            Object state = doInvoke(this.checkpoint);
+            int version = 1;
+            if (ofNullable(state.getClass().getAnnotation(Version.class)).isPresent()) {
+                version = state.getClass().getAnnotation(Version.class).value();
+            }
+            return new CheckpointState(version, state);
+        }
+        return null;
+    }
+
+    @Override
+    public boolean isCheckpointReady() {
+        boolean checked = checkpointEnabled;
+        if (shouldCheckpoint != null) {
+            checked = (Boolean) doInvoke(this.shouldCheckpoint);
+        }
+        return checked;
+    }
+
+    @Override
+    public void stop() {
+        // do we need to checkpoint here?
+        if (checkpointCallback != null) {
+            checkpointCallback.accept(getCheckpoint());
```

Comment:
What's the idea behind?
That we will update an internal state of checkpoint for connector configuration?

Reply:
yes.